src/main.rs

```diff
@@ -59,48 +64,148 @@ async fn main() {
+                    let result =
+                        match get::get_table(config.init().await.unwrap(), table, shape, format)
+                            .await
+                        {
+                            Ok(x) => x,
+                            Err(x) => format!("ERROR: {:?}", x),
+                        };
+                    Ok(result)
+                }
+                Some(("serve", _sub_matches)) => serve::app(config.init().await.unwrap()),
+                _ => Err(String::from(
+                    "Unrecognised or missing subcommand, but CGI environment vars are \
+                             undefined",
+                )),
+            },
+        };
 
     //print exit message
     match exit_result {
         Err(x) => {
-            println!("{}", x);
+            tracing::error!("{}", x);
             std::process::exit(1)
         }
 
         Ok(x) => println!("{}", x),
     }
 }
+
+#[tokio::main]
+async fn handle_cgi(vars: &HashMap<String, String>, config: &mut Config) -> Result<String, String> {
+    let shared_state = Arc::new(serve::AppState {
+        config: config.init().await.unwrap().clone(),
+    });
+    let app = build_app(shared_state);
+
+    let client = TestClient::new(app);
+
+    let request_method = vars
+        .get("REQUEST_METHOD")
+        .ok_or("No 'REQUEST_METHOD' in CGI vars".to_string())?;
+    let path_info = vars
+        .get("PATH_INFO")
+        .ok_or("No 'PATH_INFO' in CGI vars".to_string())?;
+    let query_string = vars
+        .get("QUERY_STRING")
+        .ok_or("No 'QUERY_STRING' in CGI vars".to_string())?;
+    let mut url = format!("/{}", path_info);
+    if !query_string.is_empty() {
+        url.push_str(&format!("?{}", query_string));
+    }
+    tracing::info!("In CGI mode, processing URL: {}", url);
+
+    match request_method.to_lowercase().as_str() {
+        "post" => {
+            // TODO: This is still a little bit hacky. Maybe use a library function provided
+            // by Axum (if one exists) instead of url::Url as we do below. It would also be good
+            // if we didn't have to prepend "http://example.com".
+            // Also: There may be a better way to read from STDIN using the clap library (which
+            // we are already using) instead of what we are doing below.
+            let query_str = std::io::stdin()
+                .lines()
+                .map(|l| l.unwrap())
+                .collect::<Vec<_>>()
+                .join("\n");
+            let query_str = format!("http://example.com?{}", query_str);
+            let example_url = url::Url::try_from(query_str.as_str()).unwrap();
+            let mut form = vec![];
+            for (key, val) in example_url.query_pairs() {
+                form.push((key.to_string(), val.to_string()));
+            }
+            let res = client.post(&url).form(&form).send().await;
+            let html = format!(
+                "Content-Type: text/html\nStatus: 200\n\n{}",
```

Comment:
Here and below, we need to return exactly the HTTP response headers that the client receives, not these fake headers.